api/metrics/prefix_gatherer.go

```diff
@@ -0,0 +1,96 @@
+// Copyright (C) 2019-2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package metrics
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/ava-labs/avalanchego/utils/metric"
+
+	dto "github.com/prometheus/client_model/go"
+)
+
+var (
+	_ MultiGatherer = (*prefixGatherer)(nil)
+
+	errOverlappingNamespaces = errors.New("prefix could create overlapping namespaces")
+)
+
+// NewPrefixGatherer returns a new MultiGatherer that merges metrics by adding a
+// prefix to their names.
+func NewPrefixGatherer() MultiGatherer {
+	return &prefixGatherer{}
+}
+
+type prefixGatherer struct {
+	multiGatherer
+}
+
+func (g *prefixGatherer) Register(prefix string, gatherer prometheus.Gatherer) error {
+	g.lock.Lock()
+	defer g.lock.Unlock()
+
+	for _, existingPrefix := range g.names {
+		if eitherIsPrefix(prefix, existingPrefix) {
+			return fmt.Errorf("%w: %q conflicts with %q",
+				errOverlappingNamespaces,
+				prefix,
+				existingPrefix,
+			)
+		}
+	}
```

Comment:
This is a change from the prior code.

Previously we explicitly allowed potential conflicts for metrics backwards compatibility. However, this change explicitly doesn't maintain backwards compatibility... So, I felt like it was a good time to properly restrict this. 

Previously, it would have been possible to register things like `""` and `"vm"`. If the `""` registerer added a metric that started with `"vm"`, there would have been a conflict. Now, `eitherIsPrefix` checks for this case and disallows the registration.